jax/lax_linalg.py

```diff
@@ -612,13 +612,30 @@ def _lu_jvp_rule(primals, tangents):
   u_padding[-2] = (0, n - k, 0)
   u = lax.pad(jnp.triu(lu[..., :k, :]), zero, u_padding) + u_eye
 
-  la = triangular_solve(l, x, left_side=True, transpose_a=False, lower=True,
-                        unit_diagonal=True)
-  lau = triangular_solve(u, la, left_side=False, transpose_a=False,
-                         lower=False)
+  # dummy calculation to figure out which matrices u are low-rank
+  d = triangular_solve(
+      u, jnp.ones(tuple(u.shape[:-2]) + (1, u.shape[-1]), dtype=u.dtype),
+      left_side=False, transpose_a=False, lower=False)
+  mask = (jnp.tile(jnp.isinf(d) | jnp.isnan(d),
+                   (a.ndim-2)*[1] + [a_shape[-2], 1]) &
+          ~(jnp.isinf(a) | jnp.isnan(a)))
+
+  # For non-square matrices, the shape of the mask has to be corrected
+  u_mask = lax.pad(mask, zero.astype(np.bool), u_padding)[..., :u.shape[-2], :]
+  u_safe = jnp.where(jnp.swapaxes(u_mask, -1, -2), 1.0, u)
+  la = triangular_solve(
+      l, x, left_side=True, transpose_a=False, lower=True, unit_diagonal=True)
+  lau = triangular_solve(
+      u_safe, la, left_side=False, transpose_a=False, lower=False)
+  lau = jnp.where(mask, 0.0, lau)
 
   l_dot = jnp.matmul(l, jnp.tril(lau, -1))
   u_dot = jnp.matmul(jnp.triu(lau), u)
+  # Correction for low-rank matrices
```

Comment:
This correction is not intuitive to me. Could you add a comment noting what’s going on here?

Reply:
thanks! I mostly follow this now :)

Reply:
Yeah, the full derivation is a little too long for the margin of this comment ;)